samples/compiler_plugins/xnnpack/test/fully-connected-e2e.mlir

```diff
@@ -5,16 +5,20 @@
 // RUN:     --module=- \
 // RUN:     --function=main \
 // RUN:     --input=1x2x8xi8=1 \
-// RUN:     --input=4x8xi8=2 \
-// RUN:     --input=4x8xi8=8 --xnnpack_thread_count=2 | \
+// RUN:     --input=4x8xui8=2 --xnnpack_thread_count=2 | \
 // RUN: FileCheck %s --check-prefix=CHECK-SYSTEM
 
 // CHECK-SYSTEM: EXEC @main
 // CHECK-SYSTEM: 1x2x4xf32={{\[}}[16 16 16 16][16 16 16 16]]
-func.func @main(%input : tensor<1x2x8xi8>, %kernel : tensor<4x8xi8>, %offset : tensor<4x8xi8>) -> tensor<1x2x4xf32> {
-  // TODO: Avoid doing this addition.
-  // Currently, XNNPACK will subtract the kernel by 8, so here 8 is added to the weight.
-  %kernel_adjusted = stablehlo.add %kernel, %offset : (tensor<4x8xi8>, tensor<4x8xi8>) -> tensor<4x8xi8>
+func.func @main(%input : tensor<1x2x8xi8>, %kernel : tensor<4x8xi8>) -> tensor<1x2x4xf32> {
```

Comment:
Can we try 4x8xui8? Then we might not need the type conversion on our side. If that works, we can just change the xnnpack op's signature to receive ui4 for the weight always.

Reply:
Ah, this one might be tricky if the original model source has i4 while the weight is actually ui4. I don't have a good solution for it then. We just need the hard coded conversion code anyway.

Reply:
> Can we try 4x8xui8? 

Explained in comment below

Reply:
In particular, the issue is that specifying `ui8` as the input type in `iree-run-module` does not have any effect on the actual data created for the input tensor. A `ui8=2` will pass the same data as a `u8=2`, so it does not help in this case.

Reply:
Signedness is for the most part ignored by IREE:

```mlir
func.func @main(%arg0: tensor<1x2x8xi8>, %arg1: tensor<4x8xi8>) -> tensor<1x2x4xf32> {
  %0 = stablehlo.convert %arg1 : (tensor<4x8xi8>) -> tensor<4x8xui8>
  %1 = stablehlo.convert %0 : (tensor<4x8xui8>) -> tensor<4x8xi4>
  %2 = xnnpack.fully_connected_nc_qd8_f32_qc4w %arg0, %1 : (tensor<1x2x8xi8>, tensor<4x8xi4>) -> tensor<1x2x4xf32>
  return %2 : tensor<1x2x4xf32>
}

// -----// IR Dump After ConvertStableHloToIreeInputDialects (iree-stablehlo-to-iree-input) //----- //
#executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = "default"}>
#map = affine_map<(d0, d1) -> (d0, d1)>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_embedded_elf_x86_64_]}>
module attributes {hal.device.targets = [#device_target_llvm_cpu]} {
  func.func @main(%arg0: tensor<1x2x8xi8>, %arg1: tensor<4x8xi8>) -> tensor<1x2x4xf32> {
    %0 = tensor.empty() : tensor<4x8xi8>
    %1 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%arg1 : tensor<4x8xi8>) outs(%0 : tensor<4x8xi8>) {
    ^bb0(%in: i8, %out: i8):
      linalg.yield %in : i8
    } -> tensor<4x8xi8>
    %2 = tensor.empty() : tensor<4x8xi4>
    %3 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%1 : tensor<4x8xi8>) outs(%2 : tensor<4x8xi4>) {
    ^bb0(%in: i8, %out: i4):
      %5 = arith.trunci %in : i8 to i4
      linalg.yield %5 : i4
    } -> tensor<4x8xi4>
    %4 = xnnpack.fully_connected_nc_qd8_f32_qc4w %arg0, %3 : (tensor<1x2x8xi8>, tensor<4x8xi4>) -> tensor<1x2x4xf32>
    return %4 : tensor<1x2x4xf32>
  }
}
```

Reply:
That's a bummer. Thanks for the info.